### DISCUSSION TITLE
Separate Frontmatter and Content

### DISCUSSION THREAD

**Discussion by profucius** [coddingtonbear/obsidian-local-rest-api#98] [Ideas]
Great plugin, I can really use this for a ton of use-cases.

I need the Content of an obsidian note to be separated from the Frontmatter. Currently, your plugin separates the Frontmatter, but the Content section contains plaintext of the same frontmatter text.

I imagine this would be fairly trivial to factor in, since the Frontmatter section is always preceded and followed by the standard `---` text.

I hope that you'll be able to include this. I am struggling with separating this text myself with regex, as some files frontmatter section is longer and includes different symbols than others.

**Comment by coddingtonbear** [comment]
I believe this already exists -- you just have to set your outgoing `GET` request's `Accept` header to `application/vnd.olrapi.note+json`.  If you do that, you'll receive your document formatted in a JSON format that exposes various metadata like frontmatter in fields as they're documented here: [https://coddingtonbear.github.io/obsidian-local-rest-api/#/Vault%20Files/get_vault__filename_](https://coddingtonbear.github.io/obsidian-local-rest-api/#/Vault%20Files/get_vault__filename_).

**Comment by profucius** [comment]
Thanks for the fast reply. Yes I am currently using `application/vnd.olrapi.note+json` as my accept header. I do get the Frontmatter section split in the JSON response, that is very good. However, the Content section includes the Frontmatter text. I'll paste an example below in JSON format. What I am trying to do is get _only_ the "Content text starts here" and all following text output to the Content key.

`[
  {
    "tags": [],
    "frontmatter": {
      "key1": "value 1",
      "key 2": "value 2"
    },
    "stat": {
      "ctime": 1704396610001,
      "mtime": 1704396649146,
      "size": 62
    },
    "path": "Test note.md",
    "content": "---\nkey1: value 1\nkey 2: value 2\n---\nContent text starts here."
  }
]
`

**Reply by coddingtonbear** [reply]
Oh, I see, you're not trying to _get_ the frontmatter, you're trying to get the _text_ of the document without the frontmatter.  In that case, I don't have a solution for you, unfortunately.  It seems like it should be very easy for you to do that on the receiving side, though, no, as you said?

**Reply by profucius** [reply]
That is correct. Hmm, yes I am having trouble splitting it on the receiving side. I have tried using Regex to exclude everything before the second `---` in the frontmatter. However, every file has a different count of frontmatter properties, which means an unstandardized count of newlines, until the final `---`. I am not yet seeing how to regex for a variable count of new lines between the first and last `---`. 

This is why I was hoping you may be able to incorporate this as a feature of the plugin, if you are able to already split the frontmatter out, perhaps you could exclude that frontmatter in the content too?

**Reply by coddingtonbear** [reply]
I'm afraid I really have to weigh adding extra complexity like this against how often such a feature would be useful, and I'm at the moment solidly on the side of not thinking this would be a common enough use case to merit supporting directly.

Really, I think you just need to something like:

- Check if the document's first line is `---`:
  - If it isn't:
    - There's no frontmatter; this whole document is content.
  - If there is:
     - Find the next line starting with `---`.  Your content begins on the immediately following line.

The key thing to keep in mind is that, I believe the rules for frontmatter are such that for it to exist, it must be the first content in your document, and it's delimited by two lines that contain _only_ `---` as content.  So: it should be as simple as looking for lines matching `^---$`.

**Reply by psxvoid** [reply]
Hi,

I'm trying to automate my obsidian-remote instance using n8n, and it would be very cool to have separate endpoint for read/write/add/update a specific frontmatter field.

For now, this workaround will do the job, though, it's quite a hassle:
>Really, I think you just need to something like:
>
>Check if the document's first line is ---:
>If it isn't:
>There's no frontmatter; this whole document is content.
>If there is:
>Find the next line starting with ---. Your content begins on the immediately following line.

And thanks for a great plugin, love it even in the current state!

**Reply by coddingtonbear** [reply]
I think you might be interested in https://github.com/coddingtonbear/obsidian-local-rest-api/wiki/Flexible-PATCH , @psxvoid. It's still in planning, unfortunately, and I have been extremely short on time lately so it'll be a while before it actually lands, but I do think it probably solves your problem.

**Reply by psxvoid** [reply]
> I think you might be interested in https://github.com/coddingtonbear/obsidian-local-rest-api/wiki/Flexible-PATCH , @psxvoid. It's still in planning, unfortunately, and I have been extremely short on time lately so it'll be a while before it actually lands, but I do think it probably solves your problem.

Yep, the `Flexible-PATCH` would cover my needs greatly. Totally understand the lack of free time - same for me. That's why, in THEORY, would you consider accepting a PR which might implement at least `frontmatter` manipulation part of `Flexible-PATCH`? Maybe as a separate set of endpoints, e.g. one of which might be `PUT /frontmatter/{file}` + body `{ options: { mode: "appendNewOnly" }, frontmatter: { tags: ["a", "b"], id: "my-uuid" } }`?

**Reply by coddingtonbear** [reply]
This is going to start out sounding like bad news, but I promise it's not: I'm pretty opposed to adding new routes for narrow purposes like that because once an endpoint exists, it really should exist forever, even if there's a better way of doing things that's easier to maintain going forward.

Here's the good news, though -- while I haven't had a lot of time to work on the Flexible-PATCH project, I *have* been spending a little free time day-to-day to make it possible for API extensions to be added by other plugins.  Once that's sorted out (which will probably be in the next couple days), you will be able to add new routes (like your proposed `PUT /frontmatter/{file}` one) to Obsidian Local REST API's routes via your own Obsidian plugin.

**Reply by coddingtonbear** [reply]
Try this out, @psxvoid -- https://github.com/coddingtonbear/obsidian-local-rest-api/wiki/Adding-your-own-API-Routes-via-an-Extension; just make sure you're using the latest version of Obsidian Local REST API (any version above 2.5 will be fine).

**Reply by psxvoid** [reply]
Thanks a lot! I will give it a chance. The sample looks clean and easy to understand. And the idea of plugins is brilliant as it allows to keep the core/existing set of endpoints small and maintainable while allowing other plugin developers to extend it.

**Reply by psxvoid** [reply]
@coddingtonbear Sorry for being a bit paranoid, but could you add a LICENCE to the sample repo?

**Reply by coddingtonbear** [reply]
Added!

**Reply by psxvoid** [reply]
Thanks, works as intended! And probably nobody other than me needs it, but anyway, here is just for a reference the extended API (proof of concept) with the ability to read and write frontmatter separately from a note content: https://github.com/psxvoid/obsidian-local-rest-api-extended

**Reply by profucius** [reply]
> probably nobody other than me needs it

I just got around to catching up on this thread. I wanted to say that you are not the only one, I also use n8n for projects, and originally intended to use it for interacting with Obsidian (for syncing certain data to Notion). I'm glad to see you picked this up. I'll take a look at your repo soon and see if it fits into my workflow. Edit: I see you don't have Issues/Discussions open, if I have a question later can I ask you here then? Not sure I will, just thinking ahead.

**Reply by psxvoid** [reply]
Hi Profucius,

Thanks for opening this thread in the first place. I'm also using this plugin specifically with n8n for syncing with Notion. Not ideal, because Obsidian should always be opened in the background and it eats quite a lot of memory. But it's better than nothing.

> Edit: I see you don't have Issues/Discussions open, if I have a question later can I ask you here then? Not sure I will, just thinking ahead.

Sorry, I've just used the default repository settings, that's why those sections were disabled. They should be opened now. And definitely feel free to create an issue, discussion, and even PR.
